src/ClientHandle.cpp

```diff
@@ -71,6 +71,46 @@ cClientHandle::cClientHandle(const AString & a_IPString, int a_ViewDistance) :
 	m_CurrentViewDistance(a_ViewDistance),
 	m_RequestedViewDistance(a_ViewDistance),
 	m_IPString(a_IPString),
+	m_ReceivedData(8 KiB),  // We need a larger buffer to support BungeeCord - it sends one huge packet at the start
+	m_ProtocolDataInHandler(
```

Comment:
But now you're raising protocol recogniser's logic into `cClientHandle`. Why not just have a class `cProtocolRecognizer`? You could even store it in an `std::optional<cProtocolRecognizer>` and reset it when the concrete protocol is detected. That way we could recover the memory needed for its `m_ReceivedData`.

Reply:
1. The client handle shouldn't have to carry around the baggage from an ephemeral, one-time operation for the rest of its lifetime, and even an optional still has to take up space.
2. The 'process data in' code can neatly call into whatever is needed to handle it right now, no nullptr/bool IsInUnrecognisedVersion checks leaving you wondering about whether the logic is right.
3. Responding to old versions should be some sort of gimped cProtocol that TryRecognise returns, but absent that it certainly shouldn't be something TryRecognise just happens to do, even though it's already recognised the protocol (as old) "but yeah we just have to keep on calling it in a loop to send data as a side effect", because that is surprising behaviour.

Reply:
Your implementation already carries state from `cProtocolRecognizer` in cClientHandle forever. `m_ReceivingData` and `m_ProtocolDataInHandler` only exist to serve that purpose.

If your issue is the name, then why not call it `cUnknownProtocolHandler` or something. Either way, pushing more of the logic into `cClientHandle` is just messy IMO.

Reply:
So perhaps what I'm saying is, getting cClientHandle to ...handle the state transition to SendUnrecognisedPing is a natural extension to the already-existing transition from RecogniseProtocol to SendingToProtocol in cClientHandle, because SendUnrecognisedPing is just special case of Sending with less boilerplate. It is not something that should happen in Recognise.

Reply:
oops, comment merge conflict

I intend to make cProtocol use m_ReceivingData soon, which will hopefully enable, once we have protocol stuff happening off-world, some sort of zero-copy where Recogniser and Protocol_XX use and write to the same buffer.

For ProtocolDataInHandler, well, I'd argue that this is a functional-style (?) alternative to

if (IsOldVersion)
 RespondToOld()
else if (m_Protocol != null)
 Recognise()
else
 AddReceived

so it is broader than just a way to carry state for the recogniser, but more a way to track the client handle's protocol state in its entirety.

I'll try to think about how I can put less stuff in clienthandle.